images/virtualization-artifact/pkg/controller/vmip/internal/lifecycle_handler.go

```diff
@@ -0,0 +1,147 @@
+	}
+
+	isDeletion := state.VirtualMachineIP().Current().DeletionTimestamp != nil
+	if isDeletion {
+		return reconcile.Result{}, nil
+	}
+
+	vmipStatus := state.VirtualMachineIP().Current().Status.DeepCopy()
+	vmipStatus.VirtualMachine = ""
+
+	vm, err := state.VirtualMachine(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	vmipStatus.ConflictMessage = ""
+	mgr := conditions.NewManager(vmipStatus.Conditions)
+	conditionBound := conditions.NewConditionBuilder(vmipcondition.BoundType).
+		Generation(state.VirtualMachineIP().Current().GetGeneration())
+
+	conditionAttach := conditions.NewConditionBuilder(vmipcondition.AttachedType).
+		Generation(state.VirtualMachineIP().Current().GetGeneration())
+
+	if vm != nil {
+		vmipStatus.VirtualMachine = vm.Name
+		mgr.Update(conditionAttach.Status(metav1.ConditionTrue).
+			Reason(vmipcondition.Attached).
+			Condition())
+	} else {
+		vmipStatus.VirtualMachine = ""
+		mgr.Update(conditionAttach.Status(metav1.ConditionFalse).
+			Reason(vmipcondition.VirtualMachineNotFound).
+			Condition())
+	}
+
+	lease, err := state.VirtualMachineIPLease(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	switch {
+	case lease == nil && state.VirtualMachineIP().Current().Status.Address != "":
+		if vmipStatus.Phase != virtv2.VirtualMachineIPAddressPhaseLost {
+			vmipStatus.Phase = virtv2.VirtualMachineIPAddressPhaseLost
+			mgr.Update(conditionBound.Status(metav1.ConditionFalse).
+				Reason(vmipcondition.VirtualMachineIPAddressLeaseLost).
+				Condition())
+		}
+
+	case lease == nil:
+		if vmipStatus.Phase != virtv2.VirtualMachineIPAddressPhasePending {
+			vmipStatus.Phase = virtv2.VirtualMachineIPAddressPhasePending
+			mgr.Update(conditionBound.Status(metav1.ConditionFalse).
+				Reason(vmipcondition.VirtualMachineIPAddressLeaseNotFound).
+				Condition())
+		}
+
+	case util.IsBoundLease(lease, state.VirtualMachineIP()):
+		if vmipStatus.Phase != virtv2.VirtualMachineIPAddressPhaseBound {
+			vmipStatus.Phase = virtv2.VirtualMachineIPAddressPhaseBound
+			vmipStatus.Address = common.LeaseNameToIP(lease.Name)
+			mgr.Update(conditionBound.Status(metav1.ConditionTrue).
+				Reason(vmipcondition.Bound).
+				Condition())
+		}
+
+	case lease.Status.Phase == virtv2.VirtualMachineIPAddressLeasePhaseBound:
+		if vmipStatus.Phase != virtv2.VirtualMachineIPAddressPhaseConflict {
+			vmipStatus.Phase = virtv2.VirtualMachineIPAddressPhaseConflict
+
+			// There is only one way to automatically link Ip Address in phase Conflict with recently released Lease: only with cyclic reconciliation (with an interval of N seconds).
+			// At the moment this looks redundant, so Ip Address in the phase Conflict will not be able to bind the recently released Lease.
+			// It is necessary to recreate Ip Address manually in order to link it to released Lease.
+			vmipStatus.ConflictMessage = "Lease is bounded to another VirtualMachineIP: please recreate VMIP when the lease is released"
```

Comment:
Can we move ConflictMessage to a condition message and delete this field from the vmip status?